File: src/python/hw2/Main.py
from hw2.WordMatrix import WordMatrix
import random
import codecs

def format_similar(lst):
    return ' '.join(k[1] for k in lst)

sampleWords = ['people', 'flew', 'transported', 'quickly', 'good', 'python', 'apple', 'red', 'chicago', 'language']

def eval_wordmatrix(name,matrix):
    print(name)
    for w in sampleWords:
        print(w + ':' + format_similar(matrix.eval(w, 10)))
        
    keylist = list(matrix.vWords.keys())
    for i in range(6):
        word = random.choice(keylist)
        print(word + ':' + format_similar(matrix.eval(word, 10)))
    return

v = list(codecs.open('vocab-15k.txt', encoding='utf-8').readlines())
vc10k = list(codecs.open('vocab-10k.txt', encoding='utf-8').readlines())
vc3k = list(codecs.open('vocab-3k.txt', encoding='utf-8').readlines())
vcr3k = list(codecs.open('vocab-rare3k.txt', encoding='utf-8').readlines())

inpt = codecs.open('wiki-0.1percent.txt', encoding='utf-8')

wm_10k = WordMatrix(4)
wm_10k.init(v, vc10k)
wm_10k.train(inpt)

eval_wordmatrix('Result for word count 10k', wm_10k)
pmim_10k = wm_10k.pmi()
eval_wordmatrix('Result for PPMI 10k', pmim_10k)

wm_3k = WordMatrix(4) 
wm_3k.init(v, vc3k)
wm_3k.train(inpt)
pmim_3k = wm_3k.pmi()
eval_wordmatrix('Result for PPMI 3k', pmim_3k)

wm_r3k = WordMatrix(4) 
wm_r3k.init(v, vcr3k)
wm_r3k.train(inpt)
pmim_r3k = wm_r3k.pmi()
eval_wordmatrix('Result for PPMI rare 3k', pmim_r3k)

wm_w1 = WordMatrix(1) 
wm_w1.init(v, vc10k)
wm_w1.train(inpt)
pmim_w1 = wm_w1.pmi()
eval_wordmatrix('Result for PPMI 10k, window 1', pmim_w1)

wm_w10 = WordMatrix(10) 
wm_w10.init(v, vc10k)
wm_w10.train(inpt)
pmim_w10 = wm_w10.pmi()
eval_wordmatrix('Result for PPMI 10k, window 10', pmim_w10)

File: src/java/hw2/Main.scala
package hw2

import scala.util.Random
import java.io.PrintWriter
import java.io.FileOutputStream
import scala.ref.SoftReference

object Main extends App {

  var sampleWords = List("people", "flew", "transported", "quickly", "good", "python", "apple", "red", "chicago", "language");

  def eval(title: String, out: PrintWriter, wordmatrix: WordMatrix) = {
    System.out.println(title);
    output.println(title);
    sampleWords.map(word => {
      System.out.println(word);
      out.println(word)
      wordmatrix.eval(word, 10).map(result => {
        System.out.println(result);
        out.println(result);
      });
    });

    var rlist = List[String]();
    rlist ++= wordmatrix.vwords()
    var random = new Random(System.currentTimeMillis());
    for (i <- 0 to 5) {
      var word = rlist(random.nextInt(rlist.size))
      System.out.println(word);
      out.println(word)
      wordmatrix.eval(word, 10).map(result => {
        out.println(result);
        System.out.println(result);
      });
    }
    out.flush();
  }

  var input = "res/wiki-2percent.txt"
  var output = new PrintWriter(new FileOutputStream("output"));
  /*
  var wc10k = new ArrayWordMatrix(4);
  wc10k.init("res/vocab-15k.txt", "res/vocab-10k.txt")
  wc10k.train(input)
  //eval("Result for 10K count", output, wc10k)
  var pmi10k = wc10k.pmi()
  eval("Result for 10K",output, pmi10k)
*/
  def wc3k() = {
    var wc3k = new ArrayWordMatrix(4);
    wc3k.init("res/vocab-15k.txt", "res/vocab-3k.txt")
    wc3k.train(input)
    var pmi3k = wc3k.pmi()
    eval("Result for 3K", output, pmi3k)
  }

  def wcr3k() = {
    var wcr3k = new ArrayWordMatrix(4);
    wcr3k.init("res/vocab-15k.txt", "res/vocab-rare3k.txt")
    wcr3k.train(input)
    var pmir3k = wcr3k.pmi()
    eval("Result for Rare 3K", output, pmir3k)
  }

  def wc1w() = {
    var wc1w = new ArrayWordMatrix(1);
    wc1w.init("res/vocab-15k.txt", "res/vocab-10k.txt")
    wc1w.train(input)
    var pmi1w = wc1w.pmi();
    eval("Result for Window 1", output, pmi1w)
  }

  def wc15w() = {
    var wc15w = new ArrayWordMatrix(15);
    wc15w.init("res/vocab-15k.txt", "res/vocab-10k.txt")
    wc15w.train(input)
    var pmi15w = wc15w.pmi();
    eval("Result for Window 15", output, pmi15w)
  }

  def wcnormal() = {
    var wcnormal = new ArrayWordMatrix(8);
    wcnormal.init("res/vocab-15k.txt", "res/vocab-10k.txt")
    wcnormal.train(input)
    var pminormal = wcnormal.pmi()
    eval("Result for normal", output, pminormal)
  }

  def wcweight() = {
    var wcweight = new CenterWordMatrix(8, i => { if (i > BigDecimal(8)) BigDecimal(0) else BigDecimal(8 - i) });
    wcweight.init("res/vocab-15k.txt", "res/vocab-10k.txt")
    wcweight.train(input)
    var pmiweight = wcweight.pmi()
    eval("Result for weight", output, pmiweight)
  }

  wc3k();
  wcr3k();
  wc1w();
  wc15w();
  wcnormal();
  wcweight();

  output.close();
}